diesel/src/query_builder/select_by_statement/dsl_impls.rs

```diff
@@ -0,0 +1,362 @@
+//     DB: Backend,
+//     S: QueryFragment<DB> + SelectableExpression<F> + Send + 'a,
+//     D: QueryFragment<DB> + Send + 'a,
+//     W: Into<BoxedWhereClause<'a, DB>>,
+//     O: Into<Option<Box<dyn QueryFragment<DB> + Send + 'a>>>,
+//     L: QueryFragment<DB> + Send + 'a,
+//     Of: QueryFragment<DB> + Send + 'a,
+//     G: QueryFragment<DB> + Send + 'a,
+// {
+//     type Output = BoxedSelectStatement<'a, S::SqlType, F, DB>;
+
+//     fn internal_into_boxed(self) -> Self::Output {
+//         BoxedSelectStatement::new(
+//             Box::new(self.select.0),
+//             self.from,
+//             Box::new(self.distinct),
+//             self.where_clause.into(),
+//             self.order.into(),
+//             Box::new(self.limit),
+//             Box::new(self.offset),
+//             Box::new(self.group_by),
+//         )
+//     }
+// }
+
+// impl<'a, F, D, W, O, L, Of, G, DB> BoxedDsl<'a, DB>
+//     for SelectStatement<F, DefaultSelectClause, D, W, O, L, Of, G>
+// where
+//     Self: AsQuery,
+//     DB: Backend,
+//     F: QuerySource,
+//     F::DefaultSelection: QueryFragment<DB> + Send + 'a,
+//     D: QueryFragment<DB> + Send + 'a,
+//     W: Into<BoxedWhereClause<'a, DB>>,
+//     O: Into<Option<Box<dyn QueryFragment<DB> + Send + 'a>>>,
+//     L: QueryFragment<DB> + Send + 'a,
+//     Of: QueryFragment<DB> + Send + 'a,
+//     G: QueryFragment<DB> + Send + 'a,
+// {
+//     type Output = BoxedSelectStatement<'a, <F::DefaultSelection as Expression>::SqlType, F, DB>;
+//     fn internal_into_boxed(self) -> Self::Output {
+//         BoxedSelectStatement::new(
+//             Box::new(self.from.default_selection()),
+//             self.from,
+//             Box::new(self.distinct),
+//             self.where_clause.into(),
+//             self.order.into(),
+//             Box::new(self.limit),
+//             Box::new(self.offset),
+//             Box::new(self.group_by),
+//         )
+//     }
+// }
+
+impl<S, STMT> HasTable for SelectByStatement<S, STMT>
+where
+    STMT: HasTable,
+{
+    type Table = STMT::Table;
+
+    fn table() -> Self::Table {
+        STMT::table()
+    }
+}
+
+// no IntoUpdateTarget: if you'd like to `.into_update_target`, you should not first `.select_by`
+// impl<F, W> IntoUpdateTarget for SelectStatement<F, DefaultSelectClause, NoDistinctClause, W>
+// where
+//     SelectStatement<F, DefaultSelectClause, NoDistinctClause, W>: HasTable,
+//     W: ValidWhereClause<F>,
+// {
+//     type WhereClause = W;
+
+//     fn into_update_target(self) -> UpdateTarget<Self::Table, Self::WhereClause> {
+//         UpdateTarget {
+//             table: Self::table(),
+//             where_clause: self.where_clause,
+//         }
+//     }
+// }
+
+// FIXME: Should we disable joining when `.group_by` has been called? Are there
+// any other query methods where a join no longer has the same semantics as
+// joining on just the table?
+impl<S, STMT, Rhs> JoinTo<Rhs> for SelectByStatement<S, STMT>
```

Comment:
Should we implement `JoinTo`, also `GroupBy` and `InternalJoin`, now `GroupBy` and `InternalJoin` would automatically unboxed `SelectByStatement` to `SelectStatement`

Reply:
As commented above: 
* For `InternalJoinDsl` I do not see any reason to "unbox" it
* For `GroupByDsl` I need to think about that case a bit more, especially how it interacts with the new system to handle group by clauses.

Reply:
Sorry, I just read the word `Join` not realize it is **Internal**

Reply:
What I worried about happens, I tried my best to avoid `impl AsQuery for SelectByStatement`, but `InternalJoinDsl` requires it.
Once I impl the `AsQuery` (even with where clause), the 2 impls for LoadQuery would conflict
https://github.com/diesel-rs/diesel/blob/be0854094c79ae99ba3c893242a4be358ac0b0c5/diesel/src/query_dsl/load_dsl.rs#L21-L32
https://github.com/diesel-rs/diesel/blob/e3b5faf344b004052326460b2ebafe0d0b66664a/diesel/src/query_builder/select_by_statement/mod.rs#L102-L115
I think any solution would requires a breaking change here.

Reply:
It's not clear for me why `AsQuery` is required for `InternalJoinDsl`. Could you elaborate that a bit?

Reply:
https://github.com/diesel-rs/diesel/blob/00f7de703aea02659bddad8d168b9fe6d4aee276/diesel/src/query_dsl/join_dsl.rs#L5-L11
I don't know why either, but it is required by the trait.

Reply:
I cannot answer why it's there, because removing it does not have any big influences ([That one](https://github.com/diesel-rs/diesel/blob/00f7de703aea02659bddad8d168b9fe6d4aee276/diesel/src/query_dsl/join_dsl.rs#L29) needs to be removed too, then)